src/tests/linalg/test_linalg.f90

```diff
@@ -564,18 +672,2055 @@ subroutine test_outer_product_int64
+    logical :: should_be_true_ts, should_be_false_ts, true_when_working_ts
+    logical :: true_when_working
+    write(*,*) "test_is_diagonal_rsp" 
+    A_true_s = reshape([1.,0.,0.,4.],[2,2]) !generate diagonal and non-diagonal matrices of 3 types
+    A_false_s = reshape([1.,0.,3.,4.],[2,2])
+    A_true_sf = reshape([1.,0.,0.,4.,0.,0.],[2,3])
+    A_false_sf = reshape([1.,0.,3.,4.,0.,0.],[2,3])
+    A_true_ts = reshape([1.,0.,0.,0.,5.,0.],[3,2])
+    A_false_ts = reshape([1.,0.,0.,0.,5.,6.],[3,2])
+    should_be_true_s = is_diagonal(A_true_s) !test generated matrices
+    should_be_false_s = is_diagonal(A_false_s)
+    should_be_true_sf = is_diagonal(A_true_sf)
+    should_be_false_sf = is_diagonal(A_false_sf)
+    should_be_true_ts = is_diagonal(A_true_ts)
+    should_be_false_ts = is_diagonal(A_false_ts)
+    true_when_working_s = (should_be_true_s .and. (.not. should_be_false_s)) !combine results
+    true_when_working_sf = (should_be_true_sf .and. (.not. should_be_false_sf))
+    true_when_working_ts = (should_be_true_ts .and. (.not. should_be_false_ts))
+    true_when_working = (true_when_working_s .and. true_when_working_sf .and. true_when_working_ts)
+    call check(true_when_working, &
+         msg="true_when_working failed.",warn=warn)
+  end subroutine test_is_diagonal_rsp
+
+  subroutine test_is_diagonal_rdp
+    real(dp) :: A_true_s(2,2), A_false_s(2,2) !square matrices
+    real(dp) :: A_true_sf(2,3), A_false_sf(2,3) !short and fat matrices
+    real(dp) :: A_true_ts(3,2), A_false_ts(3,2) !tall and skinny matrices
+    logical :: should_be_true_s, should_be_false_s, true_when_working_s
+    logical :: should_be_true_sf, should_be_false_sf, true_when_working_sf
+    logical :: should_be_true_ts, should_be_false_ts, true_when_working_ts
+    logical :: true_when_working
+    write(*,*) "test_is_diagonal_rdp" 
+    A_true_s = reshape([1.,0.,0.,4.],[2,2]) !generate diagonal and non-diagonal matrices of 3 types
+    A_false_s = reshape([1.,0.,3.,4.],[2,2])
+    A_true_sf = reshape([1.,0.,0.,4.,0.,0.],[2,3])
+    A_false_sf = reshape([1.,0.,3.,4.,0.,0.],[2,3])
+    A_true_ts = reshape([1.,0.,0.,0.,5.,0.],[3,2])
+    A_false_ts = reshape([1.,0.,0.,0.,5.,6.],[3,2])
+    should_be_true_s = is_diagonal(A_true_s) !test generated matrices
+    should_be_false_s = is_diagonal(A_false_s)
+    should_be_true_sf = is_diagonal(A_true_sf)
+    should_be_false_sf = is_diagonal(A_false_sf)
+    should_be_true_ts = is_diagonal(A_true_ts)
+    should_be_false_ts = is_diagonal(A_false_ts)
+    true_when_working_s = (should_be_true_s .and. (.not. should_be_false_s)) !combine results
+    true_when_working_sf = (should_be_true_sf .and. (.not. should_be_false_sf))
+    true_when_working_ts = (should_be_true_ts .and. (.not. should_be_false_ts))
+    true_when_working = (true_when_working_s .and. true_when_working_sf .and. true_when_working_ts)
+    call check(true_when_working, &
+         msg="true_when_working failed.",warn=warn)
```

Comment:
Can you please make those separate tests are at least calls separate calls to `check`?

Reply:
I think separate tests would be overkill, as you would have to repeat all the setup and teardown many many times.

Separate `check` calls should be fine though. The reason you want separate `check` calls is to determine which case(s) fail slightly more quickly and directly, right? When you say "those", I assume you want a `check` for every `should_be_*_*`, so 6 in this case?

Reply:
The latest commit includes this edit, so I'm going to resolve this for now. Feel free to change the status if needed.